### PR TITLE
Feature/destructible objects

### PR DESCRIPTION
**Pull request by KnockoutArcade**

### CODE REVIEW

File: objects/odestructibleobject/Draw_64.gml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+/// @description Insert description here
+// You can write your code in this editor
+
+draw_text(x, 64, state);
```
Comment (johnkiddjr) [inline objects/odestructibleobject/Draw_64.gml]: This appears to be debugging text that was left in, seems to draw in the middle of the life bars

File: objects/odestructibleobject/Create_0.gml [hunk 2]
```diff
@@ -0,0 +1,95 @@
+// Destructable objects act as obsticales that the player can collide with and destroy. They rely on the same
+// system that players use for hit detection, which is why a lot of the variables here are shared.
+
+hp = 50;
```
Comment (johnkiddjr) [inline objects/odestructibleobject/Create_0.gml:4]: Don't abbreviate, use ```hitPoints``` instead here.

File: objects/odestructibleobject/Create_0.gml [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+hsp = 0;
```
Comment (johnkiddjr) [inline objects/odestructibleobject/Create_0.gml:6]: Don't abbreviate, use ```horizontalSpeed``` instead here.

File: objects/odestructibleobject/Create_0.gml [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+vsp = 0;
```
Comment (johnkiddjr) [inline objects/odestructibleobject/Create_0.gml:7]: Don't abbreviate, use ```verticalSpeed``` instead here.

File: objects/odestructibleobject/Create_0.gml [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	image_xscale = 64;
+	image_yscale = 32;
+}
+hurtboxOffset = 0;
+
+// State
+state = eState.IDLE;
+
+// Variables that allow the Player to interact with this object
+#region
```
Comment (johnkiddjr) [inline objects/odestructibleobject/Create_0.gml]: Move the above comment into the region description, regions are basically a section with a comment header. Code should read like this:

```#region Variables that allow the Player to interact with this object```

File: objects/odestructibleobject/Create_0.gml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	startCombo = false;
+	cancelCombo = true; // Combo Counter shouldn't show up on Destructable Objects
+	
+	//Meter Related Variables
+	superMeter = 0; // the amount of meter the player has
+	meterBuildRate = 0.05; // The rate at which the player builds meter by approaching
+#endregion
+
+CharacterSprites = {
+	idle_Sprite : sprite_index,
```
Comment (johnkiddjr) [inline objects/odestructibleobject/Create_0.gml:88]: Use camelCase for variable naming, no underscores. This applies to the 6 lines below as well

File: objects/odestructibleobject/Create_0.gml [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+	hitByGroup = ds_list_create();
+	ds_list_clear(hitByGroup);
+	// Note to self: any time the player exits the move they are currently doing, the hitByGroup list MUST be cleared!
+
+
+	toggleIdleBlock = false;
+	cancelable = false;
+	canBlock = false;
+
+	knockbackVel = 0;
```
Comment (johnkiddjr) [inline objects/odestructibleobject/Create_0.gml:69]: Dont use abbreviations, this should be ```knockbackVelocity``` instead.

File: objects/odestructibleobject/Create_0.gml [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+	pushbackVel = 0;
```
Comment (johnkiddjr) [inline objects/odestructibleobject/Create_0.gml:70]: Ditto for ```pushbackVelocity``` here

File: objects/odestructibleobject/Create_0.gml [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+hurtboxOffset = 0;
+
+// State
+state = eState.IDLE;
+
+// Variables that allow the Player to interact with this object
+#region
+	hitstun = 0;
+	hitstopBuffer = false;
+	prevSprite = 0;
```
Comment (johnkiddjr) [inline objects/odestructibleobject/Create_0.gml:48]: Don't use abbreviations, this should be ```previousSprite``` instead.

File: objects/odestructibleobject/Step_0.gml [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+		else 
+		{
+			x = xHome - min(global.hitstop, 3);
+		}
+		
+		if (global.hitstop == 0)
+		{
+			state = eState.IDLE;
+		}
+	}
```
Comment (johnkiddjr) [inline objects/odestructibleobject/Step_0.gml:96]: This case statement is missing the break statement. Even the last case should always have a break on it.

File: objects/odestructibleobject/Step_0.gml [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+		vsp = 0;
+		if (!grounded && state != eState.LAUNCHED && state != eState.HURT && state != eState.NEUTRAL_SPECIAL && state != eState.SIDE_SPECIAL && fallDirection == 1) 
+		{
+			state = eState.IDLE;
+			grounded = true;
+			inAttackState = false;
+		}
+	}
+
+x += hsp;
```
Comment (johnkiddjr) [inline objects/odestructibleobject/Step_0.gml]: This line and the next 2 need to be indented.

File: objects/ogarbagecan/Create_0.gml [hunk 12]
```diff
@@ -0,0 +1,36 @@
+/// @description Insert description here
+// You can write your code in this editor
+
+event_inherited();
+
+hp = 30;
```
Comment (johnkiddjr) [inline objects/ogarbagecan/Create_0.gml:6]: I believe this variable is inherited, but if not please don't use abbreviations and use ```hitPoints``` instead.

File: objects/ogarbagecan/Create_0.gml [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+{
+	primary = true;
+	owner = other.id;
+	image_xscale = 21;
+	image_yscale = 32;
+}
+hurtboxOffset = -11;
+
+CharacterSprites = {
+	idle_Sprite : sprite_index,
```
Comment (johnkiddjr) [inline objects/ogarbagecan/Create_0.gml:29]: Same here, pretty sure these variable names are inherited, but if not use camelCase and no underscores, like ```hurtSprite``` etc.

File: objects/ohitbox/Step_2.gml [hunk 14]
```diff
@@ -67,7 +67,8 @@ if (collisionCheck)
 				hasHit = true;
 			} else
 			// Grabbing
-			if attackProperty.attackType[hitboxID] == eAttackType.GRAB && collision_list[| i].owner.hitstun < 1 && collision_list[| i].owner.blockstun < 1 && collision_list[| i].owner.grounded == true && collision_list[| i].owner.state != eState.THROW_TECH && collision_list[| i].owner.state != eState.JUMPSQUAT{
+			if (attackProperty.attackType[hitboxID] == eAttackType.GRAB && collision_list[| i].owner.hitstun < 1 && collision_list[| i].owner.blockstun < 1 && collision_list[| i].owner.grounded == true && collision_list[| i].owner.state != eState.THROW_TECH && collision_list[| i].owner.state != eState.JUMPSQUAT && collision_list[| i].owner.isThrowable)
```
Comment (johnkiddjr) [inline objects/ohitbox/Step_2.gml]: This if statement is getting really long, and there's a ```boolean == true``` in there... let's reformat it like this to be more readable (you should be able to basically copy paste this... spacing might be off but not by much):

```
if (attackProperty.attackType[hitboxID] == eAttackType.GRAB &&
	collision_list[| i].owner.hitstun < 1 &&
	collision_list[| i].owner.blockstun < 1 &&
	collision_list[| i].owner.grounded &&
	collision_list[| i].owner.state != eState.THROW_TECH &&
	collision_list[| i].owner.state != eState.JUMPSQUAT &&
	collision_list[| i].owner.isThrowable)
```

File: objects/ohitbox/Step_2.gml [hunk 15]
```diff
@@ -158,11 +159,18 @@ if (collisionCheck)
...
-				// Set the correct Sprite
-				collision_list[| i].owner.prevState = eState.HURT; // Set the victim's previous state to HURT
-				collision_list[| i].owner.state = eState.HITSTOP; // Set the victim's state to hitstop
-				collision_list[| i].owner.isShortHopping = false; // Make sure the victim is not using their shorthop fall speed.
 				
+				// Set the correct Sprite
+				if (!collision_list[| i].owner.isDestructibleObject) // Check if the hurbox is attatched to a destructible object
+				{
+					collision_list[| i].owner.prevState = eState.HURT; // Set the victim's previous state to HURT
+					collision_list[| i].owner.state = eState.HITSTOP; // Set the victim's state to hitstop
```
Comment (johnkiddjr) [inline objects/ohitbox/Step_2.gml]: This line can be deleted

File: objects/ohitbox/Step_2.gml [hunk 16]
```diff
@@ -158,11 +159,18 @@ if (collisionCheck)
...
+					collision_list[| i].owner.isShortHopping = false; // Make sure the victim is not using their shorthop fall speed.
+				}
+				else
+				{
+					collision_list[| i].owner.state = eState.HITSTOP;
```
Comment (johnkiddjr) [inline objects/ohitbox/Step_2.gml]: Remove this else clause, and put this line on it's own after the previous if. End result should be:

```
// Set the correct Sprite
if (!collision_list[| i].owner.isDestructibleObject) // Check if the hurbox is attatched to a destructible object
{
	collision_list[| i].owner.prevState = eState.HURT; // Set the victim's previous state to HURT
	collision_list[| i].owner.isShortHopping = false; // Make sure the victim is not using their shorthop fall speed.
}
				
collision_list[| i].owner.state = eState.HITSTOP; // Set the victim's state to hitstop
```

**Review by johnkiddjr — CHANGES_REQUESTED** [review]

File: objects/odestructibleobject/Create_0.gml (hunk 2 above)
Comment (KnockoutArcade) [reply objects/odestructibleobject/Create_0.gml:4]: The code for hitboxes checks the same variables in the destructible objects as it does in the players, so the variables must be the same name.

File: objects/odestructibleobject/Create_0.gml (hunk 3 above)
Comment (KnockoutArcade) [reply objects/odestructibleobject/Create_0.gml:6]: The code for hitboxes checks the same variables in the destructible objects as it does in the players, so the variables must be the same name.

File: objects/ogarbagecan/Create_0.gml (hunk 12 above)
Comment (KnockoutArcade) [reply objects/ogarbagecan/Create_0.gml:6]: It is, I change the value after inheritance to differentiate it from other destructible objects

File: objects/ogarbagecan/Create_0.gml (hunk 13 above)
Comment (KnockoutArcade) [reply objects/ogarbagecan/Create_0.gml:29]: The code for hitboxes checks the same variables in the destructible objects as it does in the players, so the variables must be the same name.

File: objects/ohitbox/Step_2.gml (hunk 16 above)
Comment (KnockoutArcade) [reply objects/ohitbox/Step_2.gml:172]: Destructible objects don't have a hurt state, only a hitstop state.

File: objects/odestructibleobject/Create_0.gml (hunk 4 above)
Comment (KnockoutArcade) [reply objects/odestructibleobject/Create_0.gml:7]: The code for hitboxes checks the same variables in the destructible objects as it does in the players, so the variables must be the same name.

File: objects/odestructibleobject/Create_0.gml (hunk 6 above)
Comment (KnockoutArcade) [reply objects/odestructibleobject/Create_0.gml:88]: The code for hitboxes checks the same variables in the destructible objects as it does in the players, so the variables must be the same name.

File: objects/odestructibleobject/Create_0.gml (hunk 7 above)
Comment (KnockoutArcade) [reply objects/odestructibleobject/Create_0.gml:69]: The code for hitboxes checks the same variables in the destructible objects as it does in the players, so the variables must be the same name.

File: objects/odestructibleobject/Create_0.gml (hunk 8 above)
Comment (KnockoutArcade) [reply objects/odestructibleobject/Create_0.gml:70]: The code for hitboxes checks the same variables in the destructible objects as it does in the players, so the variables must be the same name.

File: objects/odestructibleobject/Create_0.gml (hunk 9 above)
Comment (KnockoutArcade) [reply objects/odestructibleobject/Create_0.gml:48]: The code for hitboxes checks the same variables in the destructible objects as it does in the players, so the variables must be the same name.

File: objects/ohitbox/Step_2.gml (hunk 15 above)
Comment (KnockoutArcade) [reply objects/ohitbox/Step_2.gml:167]: which line?

**Review by johnkiddjr — APPROVED** [review]

**Pull request merged**